Improve support for legacy GCC versions

jhol opened the pull request:
This patch set corrects various code and build system issues encountered while building Zephyr with a legacy version of the GCC tool-chain. With these fixes, and various fixes to picolibc which have now been accepted upstream [[1]](https://github.com/picolibc/picolibc/pull/913) [[2]](https://github.com/picolibc/picolibc/pull/916), it is possible to build Zephyr on GCC version 4.1.2 and newer.

jhol committed:
cmake: compiler: Check -fdiagnostics-color=always before using

The -fdiagnostics-color flag is only on GCC versions 4.9.0 and newer.
Previously, CMake only checked that the compiler was not the XCC compiler
before using the flag.

However, given that the flag us unavailable on compilers other than XCC,
e.g. old versions of GCC, this patch replaces the previous logic with a
flag-check using the check_set_compiler_property() CMake function.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
cmake: modules/FindGnuLd: support an additional obselete version format

A legacy build of GNU ld emits the following version string:

GNU ld version 2.17.50.0.9 20070103

This patch fixes the regex parsing so that the version number is correctly
extracted from the string.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
compiler: gcc: Support versions predating include-fixed/limits.h

GCC versions starting from 4.3.0 until 12.4.0 stored the limits.h header in
a directory named include-fixed/ . Before 4.3.0 and after 12.4.0 i.e. from
13.1.0 onward, the header is stored in a directory named include/ .

Previously, target.cmake would only detect if the GCC version was 13.1.0 or
newer. This patch adds a check to also check if it is older than 4.3.0.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
include: Add support for compilers which do not define __SIZEOF_INT__

__SIZEOF_INT__ is required by zephyr_stdint.h, however the __SIZEOF_*__
macros were only introduced in GCC 4.3.

This patch adds a work-around similar to that present in picolibc's
_default_types.h which attempts to guess the correct value of the
definition based on the value of __INT_MAX__ or from INT_MAX from
POSIX <limits.h> .

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
cmake: compiler: Check -fno-printf-return-value before using

GCC versions before 7.1.0 do not have the -fno-printf-return-value flag and
give an error if it is used. This patch improves compatibility with older
compilers by making CMake check for the flag's availability before using.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
picolibc: Replace hard coded -fno-lto flag

GCC versions before 4.5 do not have the -fno-lto flag and give an error if
it is used. The flag may also be unavailable on various other non-GCC
compilers.

Previously, the picolibc CMakeList.txt script assumed that the flag was
available, and hard-coded it as an addition to the compiler flags.

This patch improves compatibility by making use of the already existing
"prohibit_lto" CMake target compiler property, which is set to the
appropriate flag to disable LTO support, or unset if the compiler lacks
such an option.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
toolchain: gcc: Added fallback for __builtin_unreachable

Support for __builtin_unreachable() was added to GCC in the 4.5.0 release.
This patch adds a fallback to __builtin_trap() for older compilers which
do not have this built-in available.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
kernel: Fixed init_delay initialization on older GCC versions

Within the K_THREAD_DEFINE macro, a nested set of macros is expanded to
populate the values of a global _static_thread_data structure.

The K_THREAD_DEFINE macro expands to statements including the
Z_THREAD_COMMON_DEFINE macro, which expands to statements including the
Z_THREAD_INITIALIZER macro, which expands to an initialization expression
of a _static_thread_data struct. The init_delay member of that struct is
then initialized with the Z_THREAD_INIT_DELAY_INITIALIZER macro, which
expands to an expression including the SYS_TIMEOUT_MS macro which
converts a number of milliseconds to a number of ticks.

For example, take the following macro:

    K_THREAD_DEFINE(
        thread_X,
        STACKSIZE,
        thread_X_entry_point, NULL, NULL, NULL,
        PRIORITY,
        0,
        0);

In abbreviated form, it expands as follows:

    typedef struct {
        int ticks;
    } k_timeout_t;

    struct _static_thread_data {
        /* ... */
        k_timeout_t init_delay;
    };

    struct _static_thread_data _k_thread_data_thread_X = {
        /* ... */
        .init_delay = (k_timeout_t){ .ticks = 0 }
    };

However, in GCC versions before 5.1, the code fails to compile with the
error "initializer element is not constant", when compiled with the
following options:

    gcc -std=gnu99 -c test.c

In the above code, the error can be corrected by replacing...

    .init_delay = (k_timeout_t){ .ticks = 0 }

...with...

    .init_delay = { .ticks = 0 }

...i.e. removing initialization with a compound literal.

In order to achieve this, this patch reworks the system of macros.

The Z_TIMEOUT_TICKS(t) macro is refactored into Z_TIMEOUT_TICKS_INIT(t)
which defines the initializer part: { .ticks = t }, and Z_TIMEOUT_TICKS(t)
which is defined as ((k_timeout_t) Z_TIMEOUT_TICKS_INIT(t)) .

For symmetry, Z_TIMEOUT_NO_WAIT_INIT is split out of Z_TIMEOUT_NO_WAIT.

Similarly, SYS_TIMEOUT_MS_INIT(ms) is split out of SYS_TIMEOUT_MS(ms) so
that the Z_THREAD_INIT_DELAY_INITIALIZER(ms) macro can use
SYS_TIMEOUT_MS_INIT(ms) which expands Z_TIMEOUT_TICKS_INIT(t) to initialize
init_delay without using a compound literal.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

andyross reviewed (dismissed):
Seems harmless.  I'm genuinely curious who out there is still using a 17 year old GPLv2 toolchain when the rest of the FSF-phobic world has long since migrated to clang.  But maybe that's not a question to ask in mixed company.

cfriedt commented:
Looks like the commit message is one character too long.

@jhol - can you amend and push?

jhol commented:
> I'm genuinely curious who out there is still using a 17 year old GPLv2 toolchain when the rest of the FSF-phobic world has long since migrated to clang. But maybe that's not a question to ask in mixed company.

You understand my struggle completely!

jhol commented:
Here is v2 of the patch-set with the following changes:

  * Check for the availability of `-fdiagnostics-color=always` on XCC in the same was every other compiler.
  * Encompassed definition of `__Z_INT_MAX` in outer check for the absence of `__SIZEOF_INT__`
  * Fixed commit message length
  * Fixed typo in commit message

tejlmand commented on lib/libc/picolibc/CMakeLists.txt:
```diff
@@ -11,7 +11,10 @@ zephyr_library_sources(
   )
 
-zephyr_library_compile_options(-fno-lto)
+zephyr_check_compiler_flag(C -fno-lto check_no_lto_result)
```
the current code on main seems to be a bad use of hard-coding `-fno-lto` as the compiler-flags has a property for exactly the reason of supporting older gcc versions / different compiler flags for this option:
https://github.com/zephyrproject-rtos/zephyr/blob/8b987d9e0293b1616ae738cdbee0bac33b71c271/cmake/compiler/gcc/compiler_flags.cmake#L25-L28

```suggestion
zephyr_library_compile_options($<TARGET_PROPERTY:compiler,prohibit_lto>)
```

Feel free to update `compiler_flags.cmake` to use `zephyr_check_compiler_flag()` instead of base the property on the compiler version.

jhol replied:
That is certainly an improvement. Thanks for the suggestion.

tejlmand reviewed (changes requested):
looks very clean to me.

Just a small request to use the `prohibit_lto` property from compiler target to follow the infrastructure design which makes it easier to handle different toolchains.

jhol commented:
Here is v3 of the patch set with the improved fix for `-fno-lto` suggested by @tejlmand

jhol commented:
@tejlmand if I understand correctly, my fix for the picolibc fork can use the same `-fno-lto` fix?: https://github.com/zephyrproject-rtos/picolibc/pull/10

tejlmand commented:
> @tejlmand if I understand correctly, my fix for the picolibc fork can use the same `-fno-lto` fix?: [zephyrproject-rtos/picolibc#10](https://github.com/zephyrproject-rtos/picolibc/pull/10)

Correct.

jhol commented:
v4 of the patch set fixes a typo in the comment in `zerphyr_stdint.h`.

Everything should be good to go now.

jhol commented:
> > @tejlmand if I understand correctly, my fix for the picolibc fork can use the same `-fno-lto` fix?: [zephyrproject-rtos/picolibc#10](https://github.com/zephyrproject-rtos/picolibc/pull/10)
> 
> Correct.

Ok, I submitted a new pull request for it here: https://github.com/picolibc/picolibc/pull/923

keith-packard commented:
That'll be upstream shortly.

carlescufi commented:
@jhol please note that I want to revivet the push for a transition to C11 or C17:  https://github.com/zephyrproject-rtos/zephyr/issues/30105. I will ping you from that issue, please take the time to explain if/why this would be a problem for you.
